components/link.py

```diff
@@ -51,12 +56,28 @@ def send(self, packet, destination, time):
             time (int):                     The time at which the packet was
                                             sent.
         """
+        destination_id = 1 if destination == self.node1 else 2
         if self.in_use:
-            # TODO: do buffer things here
-            pass
+            print "Link in use, currently sending to node %d (t = %f)" % (self.current_dir, time)
+            if self.buffer.size() >= self.buffer_size:
+                # Drop packet if buffer is full
+                print "Buffer full; packet %s dropped. (t = %f)" % (packet, time)
+                return
+            self.buffer.add_to_buffer(packet, destination_id)
         else:
-            recv_time = time + self.delay
+            transmission_delay = self.transmission_delay(packet)
+            print "Link free, sending packet %s (t = %f)" % (packet, time)
+            recv_time = time + transmission_delay + self.delay
             self.dispatch(PacketReceivedEvent(recv_time, packet, destination))
+            self.in_use = True
+            self.current_dir = destination_id
 
+            # Link will be free to send to same spot once packet has passed
+            # through fully, but not to send from the current destination until
+            # the packet has completely passed
+            self.dispatch(LinkFreeEvent(time + transmission_delay, self, destination_id))
+            self.dispatch(LinkFreeEvent(time + transmission_delay + self.delay, self, 3 - destination_id))
```

Comment:
Added a comment for `3 - destination_id`.

And the two `LinkFreeEvent` calls are because you can send in the same direction once the packet has passed completely through the opening, but you can't send in the opposite direction until the packet has completely passed through the entire link.